src/mame/drivers/dfs500.cpp

```diff
@@ -0,0 +1,1036 @@
+		, m_rombank2(*this, "rombank2")
+		, m_buzzer(*this, "buzzer")
+		, m_screen(*this, "screen")
+		, m_input1(*this, "input1")
+		, m_input2(*this, "input2")
+		, m_input3(*this, "input3")
+		, m_input4(*this, "input4")
+	{
+	}
+
+	void dfs500(machine_config &config);
+
+private:
+	virtual void machine_start() override;
+
+	void cpanelcpu_mem_map(address_map &map);
+	void cpanelcpu_io_map(address_map &map);
+	void maincpu_mem_map(address_map &map);
+	void maincpu_io_map(address_map &map);
+	void effectcpu_mem_map(address_map &map);
+	uint8_t pit_r(offs_t offset);
+	void pit_w(offs_t offset, uint8_t data);
+	uint8_t cpanel_pit_r(offs_t offset);
+	void cpanel_pit_w(offs_t offset, uint8_t data);
+	uint8_t pic_r(offs_t offset);
+	void pic_w(offs_t offset, uint8_t data);
+	void rombank1_entry_w(offs_t offset, uint8_t data);
+	void rombank2_entry_w(offs_t offset, uint8_t data);
+	void input_select_w(offs_t offset, uint8_t data);
+	uint16_t RA0_r(offs_t offset);
+	uint16_t RB0_r(offs_t offset);
+	void WA0_w(offs_t offset, uint16_t data);
+	void WB0_w(offs_t offset, uint16_t data);
+	uint16_t RA1_r(offs_t offset);
+	uint8_t RB1_r(offs_t offset);
+	uint8_t RB2_r(offs_t offset);
+	uint8_t cpanel_reg0_r(offs_t offset);
+	uint8_t cpanel_reg2_r(offs_t offset);
+	void cpanel_reg0_w(offs_t offset, uint8_t data);
+	void cpanel_reg1_w(offs_t offset, uint8_t data);
+	void cpanel_reg2_w(offs_t offset, uint8_t data);
+	uint8_t cpanel_buzzer_r(offs_t offset);
+	u32 screen_update(screen_device &screen, bitmap_rgb32 &bitmap, rectangle const &cliprect);
+	IRQ_CALLBACK_MEMBER(irq_callback);
+
+	uint8_t m_int_vector;
+	uint8_t m_sel10;
+	uint8_t m_sel32;
+	uint8_t m_sel54;
+	uint8_t m_input_sel_A;
+	uint8_t m_input_sel_B;
+	uint16_t m_maincpu_latch16;
+	uint16_t m_effectcpu_latch16;
+	bool m_TOC;
+	bool m_TOE;
+	// TODO: bool m_BVS;
+	uint8_t m_trans_rate[2];
+	uint8_t m_pattern_number[3];
+	uint8_t m_userprogram_status;
+	uint8_t m_userprogram_edit;
+	uint8_t m_trail_duration;
+	uint8_t m_snapshot;
+	uint8_t m_dot_points1;
+	uint8_t m_dot_points2;
+	bool m_buzzer_state;
+
+	required_device<v20_device> m_cpanelcpu;
+	required_device<v30_device> m_maincpu;
+	required_device<v30_device> m_effectcpu;
+	required_device<pit8254_device> m_pit;
+	required_device<pic8259_device> m_pic;
+	required_device<i8251_device> m_serial1;
+	required_device<i8251_device> m_serial2;
+	required_device<i8251_device> m_cpanel_serial;
+	required_device<pit8254_device> m_cpanel_pit;
+	required_device<upd7004_device> m_adc;
+	required_memory_bank m_rombank1;
+	required_memory_bank m_rombank2;
+	required_device<beep_device> m_buzzer;
+	required_device<screen_device> m_screen;
+	required_device<picture_image_device> m_input1;
+	required_device<picture_image_device> m_input2;
+	required_device<picture_image_device> m_input3;
+	required_device<picture_image_device> m_input4;
+};
+
+IRQ_CALLBACK_MEMBER(dfs500_state::irq_callback)
+{
+	return m_int_vector;
+}
+
+void dfs500_state::machine_start()
```

Comment:
The majority of the functionality in this function belongs in `machine_reset` instead.